command/FaviconHelper.js

```diff
@@ -0,0 +1,86 @@
+const Database = require('../data/database');
+const database = new Database();
+const url = require('url');
+const getFavicons = require('get-website-favicon');
+const scrape = require('html-metadata');
+
+class FaviconHelper {
+  constructor() {
+    this.currentRequests = {};
+  }
+
+  clearCurrentRequests() {
+    this.currentRequests = {};
+  }
+
+  getHostname(link) {
+    return url.parse(link).hostname.replace('www.', '');
+  }
+
+  async getWebsiteFavicon(link) {
+    let hostname = this.getHostname(link);
+    let data = await database.getWebsiteFavicon(hostname);
+    let expire_date = new Date();
+    // Set the expire_date to 30 days ago
+    expire_date.setDate(expire_date.getDate() - 30);
+
+    if (data && new Date(data.updated_on) > expire_date) {
+      return data;
+    } else {
+      let self = this;
+      if (!self.currentRequests[hostname]) {
+        let faviconPromise = getFavicons(link);
+        let metadataPromise = scrape(link);
+        self.currentRequests[hostname] = Promise.all([
+          faviconPromise,
+          metadataPromise
+        ])
+          .then(function(values) {
+            let publisherName = '';
+            let icon_url = '';
+
+            if (
+              values[0] &&
+              values[0].icons &&
+              values[0].icons[0] &&
+              values[0].icons[0].src
+            ) {
+              icon_url = values[0].icons[0].src;
+            }
+
+            // Picks by order of priority
+            // jsonLd name > openGraph publisher > hostname
+            if (
+              values[1] &&
+              values[1].jsonLd &&
+              values[1].jsonLd.publisher &&
+              values[1].jsonLd.publisher.name
+            ) {
+              publisherName = values[1].jsonLd.publisher.name;
+            } else if (
+              values[1] &&
+              values[1].openGraph &&
+              values[1].openGraph.site_name
+            ) {
+              publisherName = values[1].openGraph.site_name;
+            } else {
+              publisherName = self.getHostname(link);
+            }
+
+            return self.storeWebsiteFavicon(link, icon_url, publisherName);
+          })
+          .catch(function() {
+            console.log('Error occured during Favicon Promise calls.');
+          });
+      }
+      return self.currentRequests[hostname];
```

Comment:
So this is a little bit tricky. The request is complete but since other calls can be made to the same hostname, I should keep it so I don't make the same call twice (for example when title 1 is on website A and title 49 is also on website A). This is also linked to your comment on the clearCurrentRequests() call

Reply:
But I agree with you that the clearCurrentRequests() call is out of context and that everything about the currentRequests things should be in the Helper. Will try to come up with a solution

Reply:
One solution would be to remove it from currentRequest when the call to database.storeWebsiteFavicon has ended. But I'm little afraid of the concurrent executions there... If at the same time we have a database.getWebsiteFavicon and a database.storeWebsiteFavicon calls, we could have a case where getWebsiteFavicon from the Helper returns undefined.
I think the best way is to clean currentRequests when all the requests are done to avoid those kind of problems, that's why I put the cleaning in a Promise.all call, but I agree with you that it's quite a "dirty" way.